### ISSUE TITLE
hyperparams and loss function

### ISSUE THREAD

**Issue by notjedi** [Hzzone/PFA-GAN#8]
as per the paper, the loss function for the identity consistency loss should be 
<code>L<sub>ide</sub> = &lambda;<sub>ide</sub> * ((1 - &alpha;<sub>ssim</sub>) L<sub>pix</sub> + &alpha;<sub>ssim</sub> L<sub>ssim</sub> + &alpha;<sub>fea</sub> L<sub>fea</sub>)</code>, but in the actual [implementation](https://github.com/Hzzone/PFA-GAN/blob/master/models.py#L189) <code>L<sub>ide</sub></code> is as follows: `(l1_loss * (1 - alpha) + ssim_loss * alpha) * pix_loss_weight + id_loss * id_loss_weight`, which translates to <code>L<sub>ide</sub> = ((1 - &alpha;<sub>ssim</sub>) L<sub>pix</sub> + &alpha;<sub>ssim</sub> L<sub>ssim</sub>) * &lambda;<sub>ide</sub> + &alpha;<sub>fea</sub> L<sub>fea</sub></code>.

1. can you explain what is going on here?
2. what is `pix_loss_weight` and `id_loss_weight` according to the notation in the paper?
3. i tried reproducing the paper on the CACD dataset with the following hyperparams, but it doesn't seem to work (ran it for 36000 steps, but the same face reappears with no features learnt). are the hyperparams that i'm using right?
```python
init_lr = 1e-4
gan_loss_weight = 100
alpha = 0.15
pix_loss_weight = 0.02
id_loss_weight = 0.025
age_loss_weight = 0.4
decay_pix_factor = 0
```

here is the result at step 36k. i know that 36k steps is low, but there should at least be some visible changes to the face. or am i wrong?
![oBGj](https://user-images.githubusercontent.com/30691152/171530851-705e7cf5-f96c-4176-9dcd-c7d5fc58d6b3.jpg)

**Comment by BeekiG** [comment]
@notjedi
1.Do you have VGGfacedesriptor.pth and agclasfifier.pth file
2.How can I train using single GPU?
3.What will be my training file i.e .txt file or other one?
4.On dataset.py line 120 it says ValueError:empty range for randrange()(0,0,0)
Please help me and send me your code at tashe277@gmail.com

**Comment by notjedi** [comment]
1. yes, i do have them.
2. just use the command given in the README.md.
3. you need a `train.txt` file with filename, age as it contents. separated by ` `.
4. it must be because you don't have enough files in that bucket.

**Comment by BeekiG** [comment]
@notjedi 
1. So send me  both .pth files, please Email: tashe277@gmail.com
2. I have only one Gpu
3.  I think I will put my image dataset in materials and where I put the train.txt and test.txt file? for example on dataset.py line 37 and 38 data_root = osp.join(osp.dirname(osp.abspath(__file__)), 'materials', dataset_name)
    with open(osp.join(data_root, '{}.txt'.format('train' if train else 'test')), 'r') as f:
4. When you say no enough files what does it mean I don't understand sir?
 Thank you so much for your response well stay

**Comment by notjedi** [comment]
1. you can get the VGG model from the link in the paper. and as far as the age model goes, i'm afraid i can't share it. you can train your own, it barely takes a night on my 1660.
2. well, just change the `nproc_per_node` to `1` and `device_ids` to just `0`.
3. yes you put the dataset there. and `train.txt` should go to the dataset directory for example `materials/DATASET_NAME/train.txt`. same goes for `test.txt`.
4. just execute the following commands:
   - `ls materials/DATASET_NAME | shuf > materials/DATASET_NAME/train.txt`
   - `ls materials/DATASET_NAME | shuf | sed 100q > materials/DATASET_NAME/test.txt`

also, please google a few things before asking questions. or you can prolly read [this](http://www.catb.org/~esr/faqs/smart-questions.html).

**Comment by BeekiG** [comment]
@notjedi 
In models.py  there is  vgg_face.load_state_dict(load_network(osp.join('materials', 'DeepFaceVGG_RGB.pth'))) and also  age_classifier.load_state_dict(load_network(osp.join('materials',
                                                             'dex_simple_{}_{}_age_classifier.pth) so can I run the main.py without them

**Comment by Hzzone** [comment]
@notjedi 
Thanks for your interest in our work. After reading your issue, I retrained the model using the public code and checked the output results. I have updated a mistake for the pix_loss_weight while refactoring the code. There may be some gentle suggestions for you:
1. Face Alignment, as the faces you posted are not aligned by the face keypoint.
2. Longer training. I have used a small lr so it may take longer training (20w iterations in our paper).
3. Age labels. CACD dataset is noisy and not accurate, the results on MORPH may be better.

I have quickly trained on 1/5 training data, 128 image sizes with MTCNN aligned, the results are shown below:

<img width="262" alt="image" src="https://user-images.githubusercontent.com/19267349/171855148-8fb1a12f-88eb-4997-b0d0-5e682d725734.png">

The command I have used is
```
python -m torch.distributed.launch --nproc_per_node=1 --master_port=17672 main.py --image_size 128 \
--pretrained_image_size 256 --dataset_name cacd --max_iter 200000 --save_iter 100 --init_lr 1e-4 \
--gan_loss_weight 100 --alpha 0.15 --pix_loss_weight 0.002 --id_loss_weight 0.00005 \
--age_loss_weight 0.4 --name cacd
```
I note that the age model is pre-trained on the data with another alignment.

**Comment by BeekiG** [comment]
@Hzzone 
Please can you send me VGGfacedesriptor.pth and ageclasfifier.pth file? Help me sir
Email: tashe277@gmail.com

**Comment by notjedi** [comment]
thank you @Hzzone i'll look into it.

**Comment by notjedi** [comment]
@Hzzone 

1. is aligning the face really necessary? i saw another [issue](https://github.com/Hzzone/PFA-GAN/issues/4#issuecomment-877752088) where you mention that it's not compulsory to align the faces to reproduce the results of the paper.
2. also another thing that i want to understand is the <code>L<sub>ide</sub></code> in the actual implementation. you implement it as follows `(l1_loss * (1 - opt.alpha) + ssim_loss * opt.alpha) * pix_loss_weight + id_loss * opt.id_loss_weight` but it doesn't seem to be consistent with the loss mentioned in the paper, which is: <code>L<sub>ide</sub> = &lambda;<sub>ide</sub> * ((1 - &alpha;<sub>ssim</sub>) L<sub>pix</sub> + &alpha;<sub>ssim</sub> L<sub>ssim</sub> + &alpha;<sub>fea</sub> L<sub>fea</sub>)</code>
3. what is `pix_loss_weight` and `id_loss_weight` according to the notation in the paper?
4. are the hyperparams right? cause i noticed that you use `0.00005` for the `id_loss_weight` and as an effect, the model is not optimizing for the `id_loss` at all (it just keeps shooting up). this can also prolly explains the results below.

i used the hyperparams that you gave me in the previous comment and i trained on it for 33k steps and here are the results. i know that 33k steps is a little low but i should at least be seeing some meaningful changes right? or should i just train it for more epochs?

![33400_0_test](https://user-images.githubusercontent.com/30691152/172033702-e191111d-45e4-4b9f-bca6-d58fb5c3a3e2.jpg)

**Comment by Hzzone** [comment]
@BeekiG Face Alignment is necessary for many practical facial-related applications, eg, face recognition, and translation. I mean that the alignment method is not limited to Face++ in the issue you mentioned. The id loss weight can be computed by 0.02*0.025=0.0005. I used 0.00005 only for 128 image size as a quick experiment. The greater the id loss weight indicates that we need to preserve the id better.

**Comment by Hzzone** [comment]
BTW, if you wanna the output faces same as the input faces without alignment, you may need an inverse affine transformation https://github.com/XPixelGroup/BasicSR/blob/cd4e58872b3cd12ad024247165b8727de023c911/basicsr/utils/face_util.py#L157
